File: packages/core/src/pcset.ts
import type { PitchClass } from './types';
import { ALL_PITCH_CLASSES } from './types';

export function mod12(n: number): PitchClass {
  return (((n % 12) + 12) % 12) as PitchClass;
}

export function toPcSet(notes: number[]): PitchClass[] {
  const set = new Set(notes.map(mod12));
  return [...set].sort((a, b) => a - b);
}

export function transpose(pcs: PitchClass[], n: number): PitchClass[] {
  return toPcSet(pcs.map(pc => pc + n));
}

export function invert(pcs: PitchClass[], axis: number): PitchClass[] {
  return toPcSet(pcs.map(pc => axis - pc));
}

export function normalize(pcs: PitchClass[]): PitchClass[] {
  const sorted = toPcSet(pcs);
  if (sorted.length <= 1) return sorted;
  let best = sorted;
  let bestSpan = (sorted[sorted.length - 1]! - sorted[0]! + 12) % 12;
  for (let i = 1; i < sorted.length; i++) {
    const rotated = toPcSet(sorted.map(pc => pc - sorted[i]!));
    const span = (rotated[rotated.length - 1]! - rotated[0]! + 12) % 12;
    if (span < bestSpan || (span === bestSpan && lexLess(rotated, best))) {
      best = rotated;
      bestSpan = span;
    }
  }
  return best;
}

function lexLess(a: PitchClass[], b: PitchClass[]): boolean {
  for (let i = 0; i < a.length; i++) {
    if (a[i]! < b[i]!) return true;
    if (a[i]! > b[i]!) return false;
  }
  return false;
}

export function complement(pcs: PitchClass[]): PitchClass[] {
  const set = new Set(toPcSet(pcs));
  return ALL_PITCH_CLASSES.filter(pc => !set.has(pc));
}

export function areEqual(a: PitchClass[] | number[], b: PitchClass[] | number[]): boolean {
  const sa = toPcSet(a);
  const sb = toPcSet(b);
  if (sa.length !== sb.length) return false;
  return sa.every((v, i) => v === sb[i]);
}

File: packages/core-swift/Sources/ChrometriaCore/PcSet.swift
import Foundation

/// Sorted, de-duplicated pitch classes.
public func toPcSet(_ notes: [Int]) -> [PitchClass] {
    Array(Set(notes.map(mod12))).sorted()
}

public func transpose(_ pcs: [PitchClass], _ n: Int) -> [PitchClass] {
    toPcSet(pcs.map { $0 + n })
}

public func invert(_ pcs: [PitchClass], _ axis: Int) -> [PitchClass] {
    toPcSet(pcs.map { axis - $0 })
}

/// Normal form: of all rotations of the sorted set (including the identity
/// rotation), the one with the smallest span from its first to last element,
/// tie-broken by ordinary lexicographic comparison.
///
/// This mirrors `packages/core/src/pcset.ts` exactly, including its quirk
/// that the winning rotation is NOT necessarily transposed to start at 0 —
/// the untransposed `sorted` array is itself a candidate (the implicit i=0
/// case) and can win. Do not "fix" that; the vectors encode it.
public func normalize(_ pcs: [PitchClass]) -> [PitchClass] {
    let sorted = toPcSet(pcs)
    guard sorted.count > 1 else { return sorted }
    var best = sorted
    var bestSpan = mod12(sorted[sorted.count - 1] - sorted[0])
    for i in 1..<sorted.count {
        let pivot = sorted[i]
        let rotated = toPcSet(sorted.map { $0 - pivot })
        let span = mod12(rotated[rotated.count - 1] - rotated[0])
        if span < bestSpan || (span == bestSpan && lexLess(rotated, best)) {
            best = rotated
            bestSpan = span
        }
    }
    return best
}

private func lexLess(_ a: [PitchClass], _ b: [PitchClass]) -> Bool {
    for i in 0..<a.count {
        if a[i] < b[i] { return true }
        if a[i] > b[i] { return false }
    }
    return false
}

public func complement(_ pcs: [PitchClass]) -> [PitchClass] {
    let present = Set(toPcSet(pcs))
    return (0..<12).filter { !present.contains($0) }
}

public func areEqual(_ a: [PitchClass], _ b: [PitchClass]) -> Bool {
    toPcSet(a) == toPcSet(b)
}
